File: Lecture 05/Python Code/LSP/LSPViolated.py
from abc import ABC, abstractmethod

class Account(ABC):
    @abstractmethod
    def deposit(self, amount):
        pass

    @abstractmethod
    def withdraw(self, amount):
        pass

class SavingAccount(Account):
    def __init__(self):
        self.balance = 0

    def deposit(self, amount):
        self.balance += amount
        print(f"Deposited: {amount} in Savings Account. New Balance: {self.balance}")

    def withdraw(self, amount):
        if self.balance >= amount:
            self.balance -= amount
            print(f"Withdrawn: {amount} from Savings Account. New Balance: {self.balance}")
        else:
            print("Insufficient funds in Savings Account!")

class CurrentAccount(Account):
    def __init__(self):
        self.balance = 0

    def deposit(self, amount):
        self.balance += amount
        print(f"Deposited: {amount} in Current Account. New Balance: {self.balance}")

    def withdraw(self, amount):
        if self.balance >= amount:
            self.balance -= amount
            print(f"Withdrawn: {amount} from Current Account. New Balance: {self.balance}")
        else:
            print("Insufficient funds in Current Account!")

class FixedTermAccount(Account):
    def __init__(self):
        self.balance = 0

    def deposit(self, amount):
        self.balance += amount
        print(f"Deposited: {amount} in Fixed Term Account. New Balance: {self.balance}")

    def withdraw(self, amount):
        raise NotImplementedError("Withdrawal not allowed in Fixed Term Account!")

class BankClient:
    def __init__(self, accounts):
        self.accounts = accounts

    def process_transactions(self):
        for acc in self.accounts:
            acc.deposit(1000)
            try:
                acc.withdraw(500)
            except NotImplementedError as e:
                print(f"Exception: {e}")

if __name__ == "__main__":
    accounts = [SavingAccount(), CurrentAccount(), FixedTermAccount()]
    client = BankClient(accounts)
    client.process_transactions()

File: Lecture 05/Typescript/LSP/LSPViolated.ts
// Account interface
interface Account {
    deposit(amount: number): void
    withdraw(amount: number): void
}

class SavingAccount implements Account {
    private balance: number = 0

    deposit(amount: number): void {
        this.balance += amount
        console.log(`Deposited: ${amount} in Savings Account. New Balance: ${this.balance}`)
    }

    withdraw(amount: number): void {
        if (this.balance >= amount) {
            this.balance -= amount
            console.log(`Withdrawn: ${amount} from Savings Account. New Balance: ${this.balance}`)
        } else {
            console.log("Insufficient funds in Savings Account!")
        }
    }
}

class CurrentAccount implements Account {
    private balance: number = 0

    deposit(amount: number): void {
        this.balance += amount
        console.log(`Deposited: ${amount} in Current Account. New Balance: ${this.balance}`)
    }

    withdraw(amount: number): void {
        if (this.balance >= amount) {
            this.balance -= amount
            console.log(`Withdrawn: ${amount} from Current Account. New Balance: ${this.balance}`)
        } else {
            console.log("Insufficient funds in Current Account!")
        }
    }
}

class FixedTermAccount implements Account {
    private balance: number = 0

    deposit(amount: number): void {
        this.balance += amount
        console.log(`Deposited: ${amount} in Fixed Term Account. New Balance: ${this.balance}`)
    }

    withdraw(amount: number): void {
        throw new Error("Withdrawal not allowed in Fixed Term Account!")
    }
}

class BankClient {
    private accounts: Account[]

    constructor(accounts: Account[]) {
        this.accounts = accounts
    }

    processTransactions(): void {
        for (const acc of this.accounts) {
            acc.deposit(1000)

            // LSP violation: assuming all accounts support withdrawal
            try {
                acc.withdraw(500)
            } catch (e: any) {
                console.log(`Exception: ${e.message}`)
            }
        }
    }
}

// Equivalent of Java main
function main(): void {
    const accounts: Account[] = []
    accounts.push(new SavingAccount())
    accounts.push(new CurrentAccount())
    accounts.push(new FixedTermAccount())

    const client = new BankClient(accounts)
    client.processTransactions()
}

main()

/**
 * If a .ts file has no imports or exports, TypeScript treats it as a global script.
 * Adding `export {}` marks the file as a module and prevents
 * "Duplicate identifier 'SavingAccount'" errors.
 */
export {}